### PR TITLE
Adding amGithubNotifier 

### PR DESCRIPTION
**Pull request by geekodour**
continuing #208 here by replacing templates with the `description` 

Closes #104 #174 #31 #208 

This is not tested yet. will test and get put more info here.

### CODE REVIEW

File: tools/amGithubNotifier/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+A simple bridge for receiving Alertmanager alerts and posting comments to github.
+
+It listens at the `/hook` endpoint on port `:8080` by default.
+
+## Usage
+
+> Note: All alerts sent to amGithubNotifier must have the `prNum` label and `description` annotation, `owner` and `repo` labels are optional.
+
+Example `alerts.rules.yml`:
```
Comment (krasi-georgiev) [inline tools/amGithubNotifier/README.md:11]: can you put this example also in the main.go and print it with the -h flag.

Comment (geekodour) [reply tools/amGithubNotifier/README.md:11]: I think it's not possible to alter the helptext in kingpin yet, will have to add another flag for this then? https://github.com/alecthomas/kingpin/issues/119#issuecomment-207770789

I added the example in`main.go`

**Comment by geekodour** [comment]
Sample comment with this pr: https://github.com/testpromorg/prometheus/pull/1#issuecomment-516246884

**Comment by geekodour** [comment]
from #208 
> what do you mean by "multiple prombenchTests are running"? The design is that only a single test can run for each PR.

I meant multiple prombench tests running for multiple prs (one test for each pr). in #208 I mentioned that if there are multiple prs that crossed 3 days, there will be a comment on each pr with alerts of all the prrombench tests that crossed 3 days. Which is not true, because I forgot that we're grouping alerts by `namespace` and `alertname`.

So it would still work as expected, respective pr will receive respective alert. So we can close #208 and merge this pr if everything seems alright.

I branched out this pr from #208 because it had all the alert templates that might get useful sometime in the future.

cc @krasi-georgiev

**Comment by krasi-georgiev** [comment]
> - `cluster.yaml` : This is used to create the GKE cluster.
> - cluster-infra/ : These are the persistent cluster infrastructure resources.

> Create a k8s cluster -> Create the `Main node`.

	
can you change this to 
> - `cluster.yaml` : This is used to create the `Main node`.
> - cluster-infra/ : These are the persistent components of the `Main node`.


since the Schematic refer to it as the main node it would be more clear this way



> Set the following environment variables & deploy cluster. -> Set the following environment variables and deploy the cluster.

File: tools/amGithubNotifier/main.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+		g.getTargetOwner(alert), g.getTargetRepo(alert), prNum, &issueComment)
+
+	return msgBody, err
+}
+
+func (g ghWebhookReceiver) processAlerts(ctx context.Context, msg *webhook.Message) ([]string, error) {
+
+	var alertcomments []string
+
+	// each alert will have its own comment
```
Comment (krasi-georgiev) [inline tools/amGithubNotifier/main.go]:
```suggestion
	// Each alert will have its own comment.
```

Golang comments standards.

File: tools/amGithubNotifier/main.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	ctx := context.Background()
+	tc := oauth2.NewClient(ctx, ts)
+
+	return &ghWebhookReceiver{
+		ghClient: github.NewClient(tc),
+		cfg:      cfg,
+	}, nil
+}
+
+// processAlert formats and posts the alert to github
```
Comment (krasi-georgiev) [inline tools/amGithubNotifier/main.go]:
```suggestion
// processAlert formats and posts the alert to Github.
```

please check all comments everythere.

File: tools/amGithubNotifier/main.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+		      description: 'description of the alert'
+		```
+	*/
+	cfg := ghWebhookReceiverConfig{}
+
+	app := kingpin.New(filepath.Base(os.Args[0]), `alertmanager github webhook receiver
+	Example: ./amGithubNotifier --org=prometheus --repo=prometheus --port=8080`)
+	app.Flag("authfile", "path to github oauth token file").Default("/etc/github/oauth").StringVar(&cfg.authFile)
+	app.Flag("org", "default org/owner").Required().StringVar(&cfg.defaultOwner)
+	app.Flag("repo", "default repo").Required().StringVar(&cfg.defaultRepo)
```
Comment (krasi-georgiev) [inline tools/amGithubNotifier/main.go]: This is a bit confusing. Default for me is some default when not set and here is it always set from the cli arg.
the same for `org`

**Review by krasi-georgiev — COMMENTED** [review]
> - `cluster.yaml` : This is used to create the GKE cluster.
> - cluster-infra/ : These are the persistent cluster infrastructure resources.

> Create a k8s cluster -> Create the `Main node`.

	
can you change this to 
> - `cluster.yaml` : This is used to create the `Main node`.
> - cluster-infra/ : These are the persistent components of the `Main node`.


since the Schematic refer to it as the main node it would be more clear this way



few other changes


> Set the following environment variables & deploy cluster. -> Set the following environment variables and deploy the cluster.

> Setting up the test-infra -> Setup prombench

**Comment by krasi-georgiev** [comment]
can you also update the design docs with the ones from the attachment. I renamed few components.



[design.zip](https://github.com/prometheus/prombench/files/3467567/Downloads.zip)

File: README.md [hunk 5]
```diff
@@ -32,24 +32,32 @@ export GOOGLE_APPLICATION_CREDENTIALS=$AUTH_FILE
...
 > `GOOGLE_APPLICATION_CREDENTIALS` is needed for the k8s provider after [#222](https://github.com/prometheus/prombench/pull/222), long term plan is to remove this env var and pass the value of the file directly to the k8s provider.
 
-### Deploy Prometheus-Meta & Grafana
-> This is used for collecting and displaying the test results.
+### Deploy monitoring infra
+> This is used for collecting, monitoring and displaying the test results
 
 ---
 
+- Generate a GitHub auth token that will be used to authenticate when sending requests to the GitHub API.
```
[the remark this answers is not in the document]
Comment (geekodour) [reply README.md]: ~but this section is about `Deploy monitoring components` and `amGithubNotifier` is part of it and requires the `OAUTH_TOKEN`~

There's another section `Start a benchmarking test manually` it does not mention anything about Github integration.

~Mentioning it as optional will make the README instruction inconsistent. what do you think?~

File: tools/amGithubNotifier/main.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+	app.Flag("dryrun", "dry run for github api").BoolVar(&cfg.dryRun)
+
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+
+	client, err := newGhWebhookReceiver(cfg)
+	if err != nil {
+		log.Fatalf("failed to create GitHub Webhook Receiver client: %v", err)
+	}
+
+	serveWebhook(client)
```
[the remark this answers is not in the document]
Comment (geekodour) [reply tools/amGithubNotifier/main.go]: > btw wouldn't this return an error every time you hit ctrl+c ?
```
~/…/prombench/tools/amGithubNotifier
λ ./amGithubNotifier --org=a --repo=jj --authfile=auth
2019/08/08 13:42:18 finished setting up gh client. starting amGithubNotifier with a/jj
^C
~/…/prombench/tools/amGithubNotifier
λ echo $?                                                
130
~/…/prombench/tools/amGithubNotifier
λ 
```

it's exiting with `errorcode 130` this is expected right?

[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply tools/amGithubNotifier/main.go]: I still think that log.Fatal should be used only if ListenAndServe returns an error.

File: README.md [hunk 7]
```diff
@@ -7,18 +7,18 @@ It is designed to support adding more k8s providers.
 
 ## Overview of the manifest files
 The `/manifest` directory contains all the kubernetes manifest files.
-- `cluster.yaml` : This is used to create the GKE cluster.
-- `cluster-infra/` : These are the persistent cluster infrastructure resources.
+- `cluster.yaml` : This is used to create the Main Node.
+- `cluster-infra/` : These are the persistent components of the Main Node.
 - `prombench/` : These resources are created and destoryed for each prombench test.
```
Comment (krasi-georgiev) [inline README.md]:
```suggestion
- `prombench/` : These resources are created and destroyed for each prombench test.
```

File: README.md [hunk 8]
```diff
@@ -29,24 +29,32 @@ export AUTH_FILE=<path to service-account.json>
...
 ---
 
+- [Optional] If used with the Github integration generate a GitHub auth token.
+  * Login with the [Prombot account](https://github.com/prombot) and generate a [new auth token](https://github.com/settings/tokens).
+  * With permissions: `public_repo`, `read:org`, `write:discussion`.
 - Set the following environment variables
 ```
 export GCLOUD_SERVICEACCOUNT_CLIENT_EMAIL=<client-email present in service-account.json>
 export GRAFANA_ADMIN_PASSWORD=password
 export DOMAIN_NAME=prombench.prometheus.io // Can be set to any other custom domain.
```
Comment (krasi-georgiev) [inline README.md]: export DOMAIN_NAME=prombench.prometheus.io // Can be set to any other custom domain or an empty string when not used with the Github integration.

File: tools/amGithubNotifier/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (geekodour) [reply tools/amGithubNotifier/main.go]: updated.

File: README.md [hunk 9]
```diff
@@ -7,18 +7,18 @@ It is designed to support adding more k8s providers.
...
-## Setting up the test-infra
-### Create a k8s cluster
+## Setup prombench
+### Create the Main Node
 ---
 - Create a new project on Google Cloud.
 - Create a [Service Account](https://cloud.google.com/iam/docs/creating-managing-service-accounts) on GKE with role `Kubernetes Engine Service Agent` & `Kubernetes Engine Admin`. If using gcloud cli add the [`roles/container.admin`](https://cloud.google.com/kubernetes-engine/docs/how-to/iam#kubernetes-engine-roles) and [`roles/iam.serviceAccountUser`](https://cloud.google.com/kubernetes-engine/docs/how-to/iam#service_account_user) roles to the GCP serviceAccount and download the json file.
 
-- Set the following environment variables & deploy cluster.
+- Set the following environment variables and deploy the cluster.
```
Comment (krasi-georgiev) [inline README.md:21]:
```suggestion
```

File: tools/amGithubNotifier/main.go [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+		    expr: up == 0
+		    labels:
+		      severity: info
+		      prNum: '{{ $labels.prNum }}'
+		      org: prometheus
+		      repo: prombench
+		    annotations:
+			  description: 'description of the alert'
+		```
+	*/
```
Comment (krasi-georgiev) [inline tools/amGithubNotifier/main.go:67]:
```suggestion
	*/
	log.SetFlags(log.Ltime | log.Lshortfile)
```

This is to show the time and filename/line number with each log

**Pull request merged**